### ISSUE TITLE
New GUI app to visualize and edit robot-made maps (LisGein)

### ISSUE THREAD

**Issue by jlblancoc** [MRPT/GSoC2017-discussions#4]
## Initial description:
I would like to develop a new GUI application based on Qt 5.5 or next version that should be able to load data in simplemap format and visualize it as a set of different metric maps. Also it should allow editing robot poses to fix errors in automatic SLAM algorithms. I want integrate my application in cmake main project of mrpt.

See attached PDF: 
[6536204506890240_1490778297_GSoCProposaldraft.pdf](https://github.com/MRPT/GSoC2017-discussions/files/977836/6536204506890240_1490778297_GSoCProposaldraft.pdf)

**Code results**:
* [PR 527](https://github.com/MRPT/mrpt/pull/527) (~250 LOC)
* [PR 570](https://github.com/MRPT/mrpt/pull/570) (~11k LOC)
* [PR 583](https://github.com/MRPT/mrpt/pull/583) (1.2k LOC)

Video: https://youtu.be/YgGcNvrxZI0

**Comment by jlblancoc** [comment]
Student: @LisGein

Main mentor: @jlblancoc
Backup mentors: @jolting

**Comment by LisGein** [comment]
Hi @jlblancoc! Thanks so much for giving me this opportunity! I'm ready to start work. If you do not have special requests about my proposel, I'd like to start creating a prototype. First of all, I will add OpenGl to my Qt-project.

**Comment by jlblancoc** [comment]
Hi! Hope this is the beginning of a successful project! ;-)

That sounds terrific, go on for it. 

Afterwards, it would be great for you to get familiar with the [API of MRPT CMetricMaps to be rendered](http://reference.mrpt.org/devel/classmrpt_1_1maps_1_1_c_metric_map.html#a2813da9b4130aa312a002e4eb265e73b) as mrpt::opengl objects, in case you are not familiar yet... Let's keep in touch.

**Comment by jolting** [comment]
One thought I had was to accelerate the integration of cmake. I think this was in Week 9.

I think that integrating QT with the MRPT build system a little sooner might save you a lot of time in the long run. Also it makes it easier for us to build if it's part of the MRPT build process. The application should build from mrpt/apps. Furthermore, I want to emphasize code reuse. Shared GUI classes should be part of the MRPT library, not the application. All these things really requires you to integrate with the current MRPT build system as the first step.

What do you think?

P.S. Since Coding doesn't officially begin until May 30, I applaud your desire to start early.

**Comment by jlblancoc** [comment]
Can't agree more with Hunter pointers!

**Comment by LisGein** [comment]
Thank you @jolting. I think you are absolutely right. May be can I do it now?
 
And thank you @jlblancoc. I think I should read examples and MRPT code about OpenGl before include it to my project.

**Comment by jolting** [comment]
By all means, begin now.

**Comment by jlblancoc** [comment]
Hi @LisGein !

This is a general note to all accepted students that will be working on a fork of MRPT/mrpt : please, use the branch https://github.com/MRPT/mrpt/tree/mrpt-2.0-devel as **base** for all your work. In other works: **don't** start working based on `MRPT/master`, but on `MRPT/mrpt-2.0-devel`. This is because it's expected that during your project, `master` will become a maintenance branch `mrpt-1.5`, while `mrpt-2.0-devel` eventually will be merged with `master` and become the main development branch.

**Comment by jlblancoc** [comment]
@LisGein and @raghavendersahdev: Please, note that both of your projects share a first common place: integrating Qt into MRPT CMake scripts. 

So, we **strongly** recommend that both of you share your ideas, thoughts, discussions, etc. in these threads so, eventually, we can merge both of your proposed CMake changes. Recall some of the specifications: 
* Qt is optional (and the classes or apps requiring Qt should cleanly be skipped during building). 
* For Debian packaging, add the required packages in `packaging/debian/*`. This will also allow testing in the PPA or other build servers!
* For Travis and Shippable, add the required packages to the list of installed ones. 
* For Windows binary package CMake scripts: we should copy all required Qt `DLL`s alongside MRPT binaries.
Take a look at wxWidgets stuff in CMake, which just follows exactly these guidelines! ;-)

**Comment by jolting** [comment]
Take a look at how PCL does it. The MRPT cmake scripts are a little different, but the requirements are similar.
https://github.com/PointCloudLibrary/pcl/blob/master/cmake/pcl_find_qt5.cmake
Note: we do not need to support QT4 like PCL.

Make sure you look at:
http://doc.qt.io/qt-5/cmake-manual.html

Consider adding libmrptgui-qt5 to mrpt, so that Wx and Qt5 code doesn't get mixed in the same library.

For Ubuntu, you'll probably want to add libqt5opengl5-dev as a build dependency for packaging/debian/control.in and Travis/Shippable. That package depends on qt5base-dev, so it should pull in almost everything you need plus the opengl stuff.

**Comment by LisGein** [comment]
Hello! 
I started working with the branch(mrpt-2.0-devel) and noticed the compilation errors in the assimp and zlib.
They are caused by the automatic replacement of NULL with nullptr and similar changes in the two commits:
[Completely remove stlplus](https://github.com/MRPT/mrpt/commit/341c68b5667e69abb28cf67f266393fbcb97d270), [nullptr instead of NULL](https://github.com/MRPT/mrpt/commit/a4fe16ac20caf0f438859221a3527a958775193b). 

I've fixed it locally in files::
libs/base/src/compress/zlib/zlib.h
otherlibs/assimp/code/BlenderLoader.cpp
otherlibs/assimp/code/BlenderTessellator.cpp
otherlibs/assimp/code/FBXParser.cpp
otherlibs/assimp/code/XGLLoader.cpp

As I know, Assimp will soon be removed from repo, but we should fix zlib.

**Comment by LisGein** [comment]
I have forked [mrpt](https://github.com/LisGein/mrpt/tree/mrpt-2.0-devel) and added some commits. First is fixing compilation of mrpt-2.0-devel. The other things are integrating qt5 in mrpt and adding new qt application.
@raghavendersahdev, is this solution suitable?

**Comment by jolting** [comment]
You should open a pull request, so we can review it. That looks like some useful changes. Make sure you open it for the mrpt-2.0-devel branch.

**Comment by jolting** [comment]
Can you add something to script_show_final_summary.cmake?

**Comment by jolting** [comment]
Also you forgot to add script_qt.cmake to the top level CMakeLists.txt.
Finally, case the filenames like this CMainWindow.cpp and CMainWindow.h instead of all lowercase. Windows probably won't care, but Linux does.

**Comment by LisGein** [comment]
I have made a pull request, so I have fixed the compilation for Windows (checked on win 10 msvs 2017 x64 debug) and have added installation Qt dlls.

**Comment by raghavendersahdev** [comment]
Thanks @jlblancoc  @jolting  your links helped. Hi @LisGein  your solution to CMakeLists.txt looks fine to me I am also doing something very similar [here](https://github.com/raghavendersahdev/mrpt/blob/mrpt-2.0-devel/apps/benchmarking-image-features/CMakeLists.txt) . I am not using the scripts_qt.cmake currently as I am including all my dependencies in my CMakeLists, I am not sure if this is the best way, I have a CMakeLists.txt where I specify the qt, mrpt dependencies. It works fine for me.

**Comment by jlblancoc** [comment]
Hi @raghavendersahdev , 
You can now merge from the main mrpt-2.0 branch to yours, so you can use Lis's CMake scripts. Take a look at her app's CMake script. Specially, notice that find_package(), etc. are not required there, since all that part is in the other, shared, separate `.cmake` file.
Cheers!

**Comment by raghavendersahdev** [comment]
thanks @jlblancoc  I am using Lis's CMake scripts now. I have made a pull request after fixing my CMakeLists file. I am working on ubuntu 16 with Qt5

**Comment by jlblancoc** [comment]
Hi @LisGein , 

How are you doing? Any doubt on MRPT classes?

I just checked the code in your fork. Congrats, I really like your style! It's obvious you have a good background experience as a professional C++ developer :+1: 

Just some random comments: 

* No rush at all but, at some point, all .h/.cpp files should include the MRPT copyright notice. Please, add yourself to the AUTHORS file as well. 

* The "open" dialog should show both `*.simplemap` and `*.simplemap.gz`. 

* Please, add at least a brief Doxygen description to each class declaration. Not as important as if the code would go into the library (it's an app after all), but would clarify your intentions to others anyway.

* Once you start getting the feeling of having "half the work done", or earlier if you think it'll be easier like that, please consider refactoring "CGlWidget" in two: 
1) a generic renderer for MRPT COpenGLScene objects (in charge of OpenGL init, paint, mouse orbit, etc.), from which you derive: 
2) a new class, specific for your map viewer app. (Only if needed to override some special behavior)

The final goal is having the first, base class, moved to mrpt-gui, as an Qt alternative to [CMyGLCanvasBase](https://github.com/MRPT/mrpt/blob/master/libs/gui/src/CMyGLCanvasBase.cpp) which can be used in a Qt-only version of [CDisplayWindow3D](http://reference.mrpt.org/devel/classmrpt_1_1gui_1_1_c_display_window3_d.html) and obviously in other MRPT apps. Not urgent, but feels like the perfect case of code reuse... ;-)

Also: please check [here](https://github.com/MRPT/mrpt/blob/master/libs/gui/src/CMyGLCanvasBase.cpp#L80) the current standard in MRPT for mouse-based pan/orbit/zoom/... so we end up having the same behavior in either wxWidgets or Qt-based GUIs. Probably CMyGLCanvasBase and your new class in mrpt-gui could have a shared common base class, or CMyGLCanvasBase could be made compatible with both wx and Qt via 
`#if`s. Just two ideas...

Cheers

**Comment by LisGein** [comment]
Hi! Thank you :)
Yes, the last two weeks I've been working on integration Qt, mrpt and OpenGL together. I created a QtWidget that can render mrpt::opengl::COpenGLScene. 

My application now supports the following features:
- Open a dialog for simplemaps and config;
- A tree view for simplemap poses and observations;
- Create and render points and occupancy maps with opengl;
- Simple navigation (camera translate and zoom).

![image](https://user-images.githubusercontent.com/8298625/27088113-81293bc2-505f-11e7-82b0-6306ec6058ba.png)
![image](https://user-images.githubusercontent.com/8298625/27088181-a30adcbe-505f-11e7-860d-286056799106.png)


I will submit my graduate thesis next week, and after that I will be able to spend all my time to the project. 

But now, I feel that I need more files(*simplemap and *.simplemap.gz) for work. I'd like to add support of some 3D data, so I need maps with 3D laser scan. Some examples with a beacon and landmark will be good too.
Where I can found more this files?

Thank you for advises, I believe it will be useful.
I think I should further improve the navigation.

**Comment by jlblancoc** [comment]
> I will submit my graduate thesis next week, and after that I will be able to spend all my time to the project.

Good luck with that! 

On simplemaps for testing, I have collected some of them in these two Drive shared folders:

* [simplemap collection 1](https://drive.google.com/drive/folders/0B5HPWlmxIrvFOTB5MUgyeXg4VWc?usp=sharing)
* [simplemap collection 2](https://drive.google.com/drive/folders/0B5HPWlmxIrvFYkZKRE5JYmdGSUk?usp=sharing)

Most contain 2D LiDARs only, but there is at least one (named "...3lasers...") with 2D scanners pointing at different 3D orientations, which actually generates 3D pointcloud. 

Note that there are CSensoryFrames with images, stereo images, etc. I can't remember if your proposal considered this possibility, sorry if I forgot mentioning it, but definitively those images should be visible in some (optional?) panel in the UI. 

On the UI design, I understand that it's still in an early stage. I would probably recommend adding a buttons toolbar on the top with common operations, perhaps checkboxes to show/hide different layers/sensors/etc. For example, showing a little "corner" (from mrpt::opengl::stock_objects) at each sensoryframe keyframe may be one of those things one might want to show/hide.

Also, a nice feature to have is to collect all the different kinds (C++ class names) of observations in all the CSensoryFrame's of the map. Something as simple as showing this on request (via a menu or toolbar button) is enough. 

To debug the visualization of 3D range scans (e.g. Kinect, RGBD cameras) I think the best is to simple take one of the [RGBD datasets](http://www.mrpt.org/Collection_of_Kinect_RGBD_datasets_with_ground_truth_CVPR_TUM_2011) (in .rawlog format; make sure of reading the comment on 3D pointcloud not being embedded in the .rawlog), and write a very simple independent program (which does not have to be in github) that simple takes a few observations and create a .simplemap by assigning to them some 2D/3D poses. You can use arbitrary poses, or look at the ground-truth files provided in most of those datasets. 

Lots of things to do, but so far it looks great, and I'm sure you are on track...

Let us know if you need clarifications... MRPT is quite large and there are many subtleties!

**Comment by jlblancoc** [comment]
BTW: Those .simplemaps were created with MRPT 0.5.x (!!!) in 2007... It will be the definitive test for my long-term maintenance of binary serialization code. If they can be still loaded after 10 years of continuous changes it will feel fantastic !!

**Comment by jlblancoc** [comment]
**This is a kind reminder to all GSoC students and mentors**

According to [GSoC FAQ](https://developers.google.com/open-source/gsoc/faq), students are supposed to devote **30+ hours per week** to their assigned project. This amounts to 6h/day (Mon-Fri) or 4.3h/day (Mon-Sun), or any equivalent distribution given the flexibility to work on your own schedule. Clearly failing to such a commitment may be a reason to fail in the monthly evaluations, since that information was clear at the time of applying to GSoC.

It's difficult to convert hours into GitHub commits, but as a rule of dumb, there should at least a substantial daily commit during each week working day, that is, **roughly 5 substantial commits per week**. This is a general rule, and mentors may be flexible depending on the difficulty and nature of each project. 

The first evaluation will start June 26-30.

**Comment by jlblancoc** [comment]
Hi @LisGein , 

Did you manage to open those *old* simplemap files?

**Comment by jlblancoc** [comment]
For those of you creating new MRPT apps, remember to also create a basic manpage, a must-have for Debian packages. 
This includes:
* Adding one more line for your app to [doc/man-pages/CMakeLists.txt](https://github.com/MRPT/mrpt/blob/master/doc/man-pages/CMakeLists.txt), and 
* Creating a new [`.pod` file](http://perldoc.perl.org/Pod/Man.html) in [doc/man-pages/pod](https://github.com/MRPT/mrpt/tree/master/doc/man-pages/pod)

**Comment by LisGein** [comment]
Hi @jlblancoc,
Sorry I haven't been answering for too long.

Did you mean *.simplemap.gz? If yes - I did it. 

Also I added a copyright notice.  
Next I will do refactoring СGlWidget and I'll adding a brief description of Doxygen.

**Comment by jlblancoc** [comment]
Ok, good to know the files are still functional!

Btw: I think we discussed this in detail when preparing your proposal, and I'm aware it's early for this, but just in case: the goal is having the app only Asking for  .simplemap files when opening. The INI should be replaced by some GUI dialog, either with many checkboxes and fields to configure the maps, just a large edit space with the INI contents such that the user can edit, or a combination of both... 

Cheers

**Comment by LisGein** [comment]
What do you think will happen when a user opens a *.simplemap without *.ini? Nothing or will there be default settings?

**Comment by jlblancoc** [comment]
It might end up being annoying, but... I think the best is always showing the "what kind of metric maps do you need" dialog?? 
Because depending on the type of sensors stored in the .simplemap, different maps could be built.

Thinking a bit more on how this ui would be, here are some ideas:
* there should be a large text area with the ini file contents (dynamically changed and hand editable), which will be the text that ultimately gets loaded in a CMemoryConfigFile and passed to the metric map when clicked OK.
* a button for loading from an ini file, as now, but then its contents gets loaded in the text field and there is a chance to edit it before building maps.
* a couple buttons for "typical  setups", e.g. only pointcloud, only gridmap,...

The CMemoryConfigFile class would allow you to easy read and write specific ini file entries by loading the text area text into it, editing, then saving back as a string.

What do you think?

**Comment by LisGein** [comment]
Perhaps it will be good if the program alternately tries to open maps (Points map, Beacon, e.t.c) until one of them opens up?

"typical setups" -- what do you mean?

**Comment by jolting** [comment]
Hmmm. Maybe something like the way rviz works. Each type of data that can be visualized is listed and can be added to the the visualizer. I like how there is an easy way to add each type of data with rviz. Also I like how there is a menu that lets me tweak how each visualizable element is displays. Most importantly, I can easily turn them off when I'm not trying to visualize them.

**Comment by LisGein** [comment]
Thanks for the advice! So, I have started working with the configuration widget.

**Comment by jlblancoc** [comment]
Good idea, Hunter!

As a first approach, the configuration of how each metric map is
built, and rendered, is encoded in its "Initilization" section of the
configuration ("INI") file; so the configuration UI for each metric
map could be the same / based on / etc. the one we discussed
previously, Alice, regarding the questions the user is asked when a
simplemap is opened.

Search for `MAP_DEFINITION_START` and `MAP_DEFINITION_END` in all maps
headers to see the list of parameters for each map.

**Comment by LisGein** [comment]
What do you mean the initialization section? As I can see, for example CSimplePointsMap loads from config CPointsMap::TInsertionOptions and CPointsMap::TLikelihoodOptions. Where is the common part?

**Comment by LisGein** [comment]
What about evaluation? Should I press the button right now?

**Comment by LisGein** [comment]
Hi! 
I did a widget for configuration. It can modify, add and remove configs for metric maps. Also it can load from *.ini file. 
![image](https://user-images.githubusercontent.com/8298625/27771409-8e35b362-5f56-11e7-92d5-ac8e2c9ccd67.png)

I tried to save the configuration(TSetOfMetricMapInitializers::dumpToTextStream) in a file, but this function uses different format. For example: "----------- [CBeaconMap::TInsertionOptions] ------------"
but needed "[MappingApplication_beaconMap_00_insertOpts]". Does MRPT contain the possibility of saving in the correct format?

Also I need to specify a range for each numeric control. Is it documented somewhere?

**Comment by jlblancoc** [comment]
Hi Lis! 

Well, that's UI is great, thanks for the effort.... definitively, you started with one of the more complex option structures!

Your question on ranges, together with me thinking of how much work may it be to MAINTAIN your UIs in the future (imagine those parameters changing in the future!!) made me reconsider this part of your project... 

I think it's time to address this in a "Proper Way (TM)" ;-)

What about adding [reflection](https://en.wikipedia.org/wiki/Reflection_(computer_programming)) to MRPT `CLoadableOptions`, @LisGein , @jolting ? 
I thought about a way to do it that would not be really that hard: we already have all CLoadabaleOptions equipped with save and load methods. If we extend the "save" methods [here](https://github.com/MRPT/mrpt/blob/master/libs/base/include/mrpt/utils/CConfigFileBase.h#L60) such that CConfigFileBase keeps an internal map of variables to types (more discussion on "types" required....), we could just create any CLoadableOptions structure, let it be its default values, then "save" (into a dummy memory config file, for example), and we will have the list of all variables and their types. 
Via TEnumType we have access to values of enum types, conversion of numerical values to their corresponding strings, etc.

If we want to extend this with valid value ranges and such, we'll need to modify all existing CLoadableOption classes, but if we leave that as optional, we would have, right now, without any important code change elsewhere, the list of data members and their type... enough for dynamically generating the UIs for **any** CLoadableOptions on the fly. It would be **great**, especially thinking of maintainability :-)

Thoughts?

@LisGein : On your last question, I think I don't get it, sorry... Each map has a set of parameters for:
* 1) construction, see [this](http://reference.mrpt.org/devel/structmrpt_1_1maps_1_1_t_metric_map_initializer.html), 
* 2) most of them also have a likelihood options (which are not actually relevant for **map construction**, so they may be not 100% required for this app... unless we add new functionality in the future, so I don't discard allowing users to change it too), and 
* 3) parameters for "insertion" of observations (this one is required for this app).

Let me know if this helps to your doubt...

Cheers

**Comment by LisGein** [comment]
Hi!
I agree that a more general way to viewing and editing properties is a good idea. But the current solution is simple and works. So I think that can be done after the other main features from my proposal will be implemented.

Cheers

**Comment by jlblancoc** [comment]
Sure! Firstly, focus on the rendering of all kind of maps, mouse operations
like selecting a pose, etc

Let me know if you have a problem with anything of this, so you don't waste
time looking for mrpt documentation.

**Comment by jlblancoc** [comment]
Hi @LisGein ! 

I browsed your code and all I see is fine... I like you already addressed the re-factorization of the wx vs. Qt OpenGL widgets. Good!

Just one question: aren't these two sets of parameters duplicated? 
* https://github.com/LisGein/mrpt/commit/ffe543c3167c89e87e4cd0e951ffda1c34201ef6#diff-73b633cc8f356993db9aff01605d02fbR28
* https://github.com/LisGein/mrpt/commit/ffe543c3167c89e87e4cd0e951ffda1c34201ef6#diff-73b633cc8f356993db9aff01605d02fbR28

**Comment by LisGein** [comment]
Hi @jlblancoc! 
Yes, I thought to stay only the CamaraParams structure, but for this it is necessary in the whole project to make use the functions of getters. 
What do you think about it?

**Comment by jlblancoc** [comment]
Since we are aiming at the 2.0 version of mrpt... it may be
acceptable, if in that way all the camera parameters are more
centralized and accessed via getter/setters instead of being plainly
exposed...

**Comment by LisGein** [comment]
Hi!
Before the pull request I will add getters and setters in the CGlCanvasBase.

I have added:
* visualize all robot poses on maps(red points):
![image](https://user-images.githubusercontent.com/8298625/28134475-bb691ff2-674b-11e7-9c4f-4a78ec6f520e.png)

* opening stereo images and image:
![image](https://user-images.githubusercontent.com/8298625/28134771-9fc36608-674c-11e7-959b-201e050e6078.png)


* visualize robot direction and laser scan:
![image](https://user-images.githubusercontent.com/8298625/28134683-62759fdc-674c-11e7-9e46-a5adee89e077.png)
(I liked the giraffe :D )

**Comment by jlblancoc** [comment]
Great progress Alice!

I would only recommend changing the dots of poses to another visualization
that shows the heading. The simplest is
mrpt::opengl::stock_objects::simplexyzcorner.

Note that pull requests should go now against master. Then we'll probably
have to pass the new clang-format parser to your code, so it would save
time and problems if you make your code conform to the new clang format
file that you can find at the root directory in master.

Cheers

**Comment by LisGein** [comment]
Hello!

I changed the dots. 

I have a few questions: 
 * I think for 3D maps doesn't need minimap. Can I don't make this?
 * I wrote in my proposal:
      "8.2 Software should have tools for editing robot poses:
               8.2.1 Adding on a map."
    But now I think that it's useless. What do you think about it?
 *  You said, "pull requests should go now against master". 
    What is the right way to do this with git?

Cheers

**Comment by jlblancoc** [comment]
> I think for 3D maps doesn't need minimap. Can I don't make this?

Ok, it's fine!

> I wrote in my proposal:
> "8.2 Software should have tools for editing robot poses:
> 8.2.1 Adding on a map."

Adding is not necessary, ok. Changing the pose, that **is** useful ;-)

> You said, "pull requests should go now against master".
> What is the right way to do this with git?

Using the GitHub web interface, it's super easy: just view your branch online and clicking on "pull request", it will automatically propose you to open the PR against `master`.

**Comment by LisGein** [comment]
When should I do pull requests?

I added several features:
* Enable/disable the grid;
* Change the background color; 
* Change the grid color; 
* Shortcuts for open simplemap and for open configuration;
* Select the pose. 

Now, I am making a multi selection(with a rectangle and with a ctrl). 

Can you check the code in [CGlWidget::mousePressEvent](https://github.com/LisGein/mrpt/blob/mrpt-2.0-devel/apps/robot-map-gui/gui/glWidget/CGLWidget.cpp#L359)? 
When I click a pose in the corner of the scene, I see a small error(about 10 pixels).
How can I fix it?

**Comment by jlblancoc** [comment]
Good amount of work!

In your case, I think it makes sense to open a PR against master now, so we
see the output of travis and such.

I'll take a look at the code you mention...

**Comment by LisGein** [comment]
Hi! 
Sorry for the long silence.
I have done multi selection with ctrl and I added in the general settings setting for robot poses and selected robot poses(size and colour).
Now, I'm preparing a PR. Today I am doing it.  

Cheers :)

**Comment by jlblancoc** [comment]
Great, I'm waiting for it!

**Comment by LisGein** [comment]
I did it, but this branch has a lot of conflicts. 
Is it normal?

**Comment by jlblancoc** [comment]
Ohh! You broke mrpt! 😢




Just kidding! ;-)
Normally, no conflicts are expected, but given that these months we have
been doing an overall redesign of the entire mrpt towards the new major
release 2.0... yes, I expected conflicts.

Not your fault, will try to fix them myself, no worries.

**Comment by jlblancoc** [comment]
Hi @LisGein ! 

I just managed to solved all the conflicts after the merge and could successfully test your app... great progress! :+1: 

However, now that the app starts to approach a more definitive state, I feel it's time to make some improvements (put them in your list and process them in the order you prefer): 
* Just a matter of consistency: locale LC_NUMERIC is used by wxWidgets and Qt to select the format of numbers depending on the country. For all MRPT wxWidgets apps however, I have overridden it manually to use the "C" numeric format (e.g. decimal separator is ".", no thousands separator, etc.). I think it's better like that so it's the same on all computers, easier to copy and paste, we avoid problems reading from config files, etc. Please, investigate how to do the same with Qt... hope it's easy! You can take a look at the first lines of any wxApp program to see how it's done in wxWidgets.
* GridMap params, Creation tab: resolution spin button moves in steps of 1 meter (!). 0.01 would be better.
* If you add one or more metric maps, then remove them, the last tab remains there. Shouldn't be deleted?
* The "About" button is not implemented. It's great you didn't, because it was recently refactored for mrpt-2.0. You can see how it's done in the rest of apps. However, notice it's done with **wxWidgets**... I'm sure you would get the point and add the equivalent for Qt without too much effort. 
* I would add an "Exit" menuitem in "File".
* Also, it would be great to add all operations that right now are available via Menu items, as toolbar buttons / checkboxes. 
* When the GUI is opened for the first time, first-time users may get confused about how to use this. A text area with short instructions would be fantastic. Perhaps as a tab in the same space that later will be occupied by the metric maps. Any other alternative idea? Also, once a simple map is loaded, the dialog shown with the "add" button should be automatically invoked, since a map without a metric map is (almost?) useless... 
* Robot poses are correctly represented as oriented XYZ corners **after** they are clicked. It's more informative to *always* display them as XYZ corners instead of a point cloud. This will imply changes in the display configuration UI, of course...
* I feel automatically enabling the "show configuration" (the side panel) by default is a good idea.
* If one adds two occupancy grids, they are named "...0" and "...1" in the tabs, but not in the configuration list (!). Hopefully it's easy to fix.
* A "recent files" menu entry under "File" is another useful thing to have, which is present in most other MRPT GUIs.
* If one clicks on a pose, it gets selected. But then, clicking around it seems to have the effect of "moving" the pose (!!). Is this intentional? In any case, the actual pose stored in the simple map is not affected, since pressing "apply" in the gridmap config pane, does not change the built metric map. At least, that's the way I would expect a modified map to be regenerated... any comment/hint on this? 

That's all for now! :-)  Despite the long list, it's plainly obvious you did an amazing job! 
Naturally we'll fill the evaluation without waiting for you to get hands on these tasks, so take the time you need to process them and continuing with your other planned tasks.

**Comment by jlblancoc** [comment]
PS: I've started to update the first comment on the top, to reflect the list of your PRs. It's useful if you want to submit the URL of this issue ticket as a "results" link at the end of GSoC. Making a YouTube video is also an ideal final add-on, but obviously that will have to wait a bit yet...

**Comment by LisGein** [comment]
Thank you :)

*  You wrote: "locale LC_NUMERIC is used by wxWidgets and Qt to select the format of numbers depending on the country" 
Can you give an example this numbers? I've never seen this.

* About steps of the resolution spin button: I stopped this task until reflections are not ready.

* The last metric map is not deleted: Yes, it's a bug. I'll fix it. Thank you!

* I want to implement the "About" button and "recent files" later. It's really very useful.

* "Exit" in "File": no problem :)

*  Add all operations as toolbar buttons/checkboxes: Do you mean "open simplemap", "show observations tree", etc?

* When the GUI is opened: Yes, it's a great idea. I would like to develop it after tasks in my plan.

* Robot poses as oriented XYZ corners: I'll do it. 

* automatically enabling the "show configuration": Yes, I want to do this, but now this hinders. 

*  A metric map named "...0" and "...1" in the tabs, but not in the configuration list: Yes, thank you. It was a bad refactoring. I'll fix it.

* Clicking around the robot pose seems to have the effect of "moving" the pose:
:D It was a bad idea use the position of the clicked pose the mouse cursor. I'm sorry, I'll fix it, too.

**Comment by jolting** [comment]
The setlocale thing is also part of the C standard and is used by many of the print functions.
http://en.cppreference.com/w/cpp/locale/setlocale

LC_NUMERIC specifies the decimal point, the thousands seperator and the grouping. In the US we use a decimal point. Many countries use decimal commas.

**Comment by jlblancoc** [comment]
@LisGein On `#include <QGLWidget>`, [I read](https://stackoverflow.com/a/37964784/1631514) that: "As suggested in the Qt documentation on QGLWidget, starting from Qt 5.4 you should use QOpenGLWidget instead.". 

Probably not critical...

**Comment by jlblancoc** [comment]
> You wrote: "locale LC_NUMERIC is used by wxWidgets and Qt to select the format of numbers depending on the country"
> Can you give an example this numbers? I've never seen this.

As Hunter said above, in many countries (including here in Spain), decimal separator is a ",". So, when showing CPose2D poses, I see "[1,200 , 4,500 , 6,700]"  instead of "[1.200 4.500 6.700]". This, together to the fact that we often format numbers with three digits, plus that the thousands separator is a "."... is **extremely** annoying and prone to errors! Also, saving numbers in that format is not computer-independent, etc.

Even if one sets language to English during the OS installation, as I do (GCC build errors in Spanish drive me mad!), the installer use the national standards for other things (run "locale" on a console to see it), apparently based on the time-zone selection.

So, please, select the "C" locale using setlocale() to see if that fixes the issue. You could probably test it by manually executing `export LC_NUMERIC=es_ES` on a console, then running your app. You should see "," as decimal separator. In this way you could see if the program explicitly sets the "C" locale correctly.


> Add all operations as toolbar buttons/checkboxes: Do you mean "open simplemap", "show observations tree", etc?

Yep!

**Comment by jlblancoc** [comment]
One more issue: if one moves the zoom slider at the bottom, right after opening the app (with no active metric map yet), the program crashes with seg. fault (!!). 

Hopefully it's just a missing `if()` check... 

I have also verified that, at least in Linux, the non-Qt CMyGLCanvas is not working after the refactoring :-( 
You can see it, for example, in the app `SceneViewer3D`.

**Comment by LisGein** [comment]
About locale: I understood, thanks! I'll do it.

@jlblancoc thanks, for the note of QOpenGLWidget. I didn't know. I'll change it. 

I'll see CMyGLCanvas. 
I want to suggest to change the class name to CWxGlCanvas.
Also, I found a small mistake in the base class(I added check min and max).

missing if(): Yes, thank you for excellent testing! I'll fix all bugs.

"GCC build errors in Spanish drive me mad!" -- yes, I use English instead of Russian, too.

**Comment by LisGein** [comment]
I want an advice. How do you think can I delete my MRPT's fork and fork again?

**Comment by jlblancoc** [comment]
In the past I have done that by deleting my fork and creating a new one
again... Not too elegant, but fast!

**Comment by LisGein** [comment]
Thank you! 
What is the elegant way?

**Comment by jlblancoc** [comment]
I guess doing it via a merge and/or rebase is less dramatic, but I'm
actually not sure if it will add unnecessary commits to the git tree...

**Comment by LisGein** [comment]
I have a compilation error in otherlibs/assimp/code/ACLoader.cpp on line 670 
boost::scoped_ptr<Subdivider> div(std::make_aligned_shared<Subdivider>(Subdivider::CATMULL_CLARKE));

May it's mean mrpt instead of std?

**Comment by jlblancoc** [comment]
Just set BUILD_ASSIMP=Off for now...

**Comment by jolting** [comment]
Assimp is unnecessary, but assimp uses boost pointers. That needs to go back to boost::make_shared

**Comment by LisGein** [comment]
@jlblancoc  thank you! I forgot it.

@jolting  do I need to do this?

**Comment by jolting** [comment]
No. This wasn't your fault. If you'd like to fix it go ahead.

**Comment by jolting** [comment]
Also, here's my input on the earlier conversation. The difference between merge and rebase is that merge isn't suppose to rewrite git history. The idea of "reforking" a branch is essentially rebasing.

A merged keeps all the patches applied to the old branch and creates a merge commit that handles the conflicts. A rebase reapplies all the patches that were formerly applied to an old branch, but applies them to a new branch. That means that resolving a conflicts involve rewritting the individual commits. Even if there aren't any conflicts, this does rewrite git history because applying a commit on a different base creates a different hashes.

Rewriting public Git history is generally considered a bad idea. If someone pulls your git history and then works off of it and you rewrite your history then there will be a mismatch when it comes to git hashes and syncing back up will be difficult.

GitHub somewhat changes the model for git that is commonly used in other parts of the open source world. Many people publish their incremental work on GitHub. Git history should be clean. One approach is never push anything to github unless you intend someone else to pull it. I find this a bit too dogmatic.

I think it's generally accepted that the main MRPT repository will never rewrite Git history unless absolutely necessary. You are free to rewrite your Git history any way you wish on your own fork. Rebase is the tool to do that.

You can simply use the --rebase option when doing the `git pull`. By default git pull is like a meta command that does a `git fetch` and a `git merge`. The `--rebase` does the `git fetch` and then does a `git rebase`.

**Comment by jolting** [comment]
One more thing. If you do a rebase, then you will have to force the push when pushing to github because github will complain about the mismatching git history.

**Comment by jlblancoc** [comment]
Hi! @jolting Of course, I totally agree with you in that rebasing is bad practice for public repositories. 
However... being pragmatic, doing it in a "private" branch or in a fork, when you are "sure" that the rebase will not affect anyone else should not be just "okay"? :-)

@LisGein :
> I'll see CMyGLCanvas.
> I want to suggest to change the class name to CWxGlCanvas.

Sure. I guessed you meant `CMyGLCanvasBase`, right? 
Done here: https://github.com/MRPT/mrpt/commit/12886ca01b644f1c318b12d816ed3918daccfcfe

**Comment by jlblancoc** [comment]
Hi again @LisGein !

Do you have access to a Windows? I tested today your app with Visual Studio 2015 + Qt5.7 and I needed to make a number of changes to make it build. Nothing serious, just take a look [here](https://github.com/MRPT/mrpt/commit/8bbe6c27c658ad71851c4034911682b8ed243934) and [here](https://github.com/MRPT/mrpt/commit/feacdfa251420909b5f0e7714ac5996c03ffbee4) to be aware of them. There might be more errors to solve due to missing `#if` guards; recall that Qt is an **optional** dependency to MRPT, so everything must build cleanly without it.


The bad part is: building in debug and trying to load a .simplemap led to a **Qt assertion and a crash**. Perhaps it's because I'm using (temporarily) an MRPT configuration without OpenCV, but I'm not sure the error is related to this. I attach a snapshot just in case it helps you to identify the problem. 

![captura de pantalla 2017-07-30 02 35 01](https://user-images.githubusercontent.com/5497818/28749691-8b207240-74d0-11e7-9de8-8a563fbc6115.png)


Debugging with Visual Studio, IMO, is a smooth experience, easier than in GNU/Linux (again, IMHO!!). So having these errors it's super easy to click "debug" and inspect and play around the code and the call stack and such.

**Comment by LisGein** [comment]
@jolting thank you for the explanation! It's useful information for me.

@jlblancoc I have VirtualBox. I want to try to repeat it. 
How it's possible? (I tried with enabled in cmake-gui DISABLE_OPENCV, but I think I forgot something because it doesn't compile).

I usually use Linux and Windows without a console looks awful :D

**Comment by LisGein** [comment]
I have done:
* locale;
* fixed deleting a last metric map form viewer;
* added to metric maps name numbers in tabs;
* fixed crash when nothing opened and using zoom controller;
* added "Exit" in "File";
* fixed the selected "dancing point";
* replaced QGLWidget to QOpenGLWidget;
* default m_minZoom == -1.0f, m_maxZoom == -1.0f in CGlCanvasBase(because In the previous base class this was not)

**Comment by jolting** [comment]
>doing it in a "private" branch or in a fork, when you are "sure" that the rebase will not affect anyone else should not be just "okay"? :-)

I think it's okay. That goes for private(payed for or educational) or personal(public) Git repositories.

`git log --graph` will give you a good idea about the complexity of the Git History. If it looks like spaghetti you may be doing something wrong.

**Comment by LisGein** [comment]
Hi! 
I have done deleting observations from maps and undo/redo. 
The next thing I will do is update the map after deleting(now I create new widgets) and deleting robot poses from observations tree.  

Thanks to @jlblancoc I found out information about const references in Qt signals and slots. Also, I found out information about [New Signal Slot Syntax](https://wiki.qt.io/New_Signal_Slot_Syntax).

**Comment by jlblancoc** [comment]
> Thanks to @jlblancoc I found out information about const references in Qt signals and slots. Also, I found out information about New Signal Slot Syntax.

👍 cool to learn on that new syntax, thanks! 

> I have done deleting observations from maps and undo/redo.

Will check it asap and give you some feedback.

**Comment by jolting** [comment]
Neat. I've found that macros don't play nicely with type deduction or at least they are unintuitive.

I can't wait until c++ is advanced enough to eliminate the need for Qt moc altogether. Without resorting to uglier macros, there are still some c++ features required before moc can be eliminated.

**Comment by LisGein** [comment]
Hi!
I have done moving observations from maps and undo/redo for this. Also, I have added the toolbar for operations from the menu and icons for almost each icon :)

**Comment by jlblancoc** [comment]
Great! I see you have been adding more features these last days. Will test the app and come back if I find something worth mentioning... ;-)

**Comment by jlblancoc** [comment]
Some further points that would be great to have: 

* Use `AppStartMenuLink` as [in here]( https://github.com/MRPT/mrpt/blob/master/apps/RawLogViewer/CMakeLists.txt#L61) to add a link to the GUI app in Windows binary packages. 

* For Debian packages, it would be also great to do the same that we have now with ".rawlog" files: after installing `mrpt-apps` one can simply double-click on a .rawlog file and it will be opened with RawLogViewer. For that to happen, there are a number of steps:
  * Define `.simplemap` (and .simplemap.gz) as a special kind of MIME type. See [this](https://github.com/MRPT/mrpt/tree/master/share/mime/packages). Use, for example `x-mrpt-simplemap` as mime type.
  * Add a `.desktop` file. Copy the [file for RawLogViewer](https://github.com/MRPT/mrpt/blob/master/share/applications/rawlogviewer.desktop) as a template. See that the mime type here must match the one we defined above. 
  * The icon used in the last point must be [added here](https://github.com/MRPT/mrpt/tree/master/share/pixmaps).
  * Add an [appdata](https://github.com/MRPT/mrpt/blob/master/share/appdata/rawlogviewer.appdata.xml) file, which might be used by software centers, etc. Let me know what screenshot you want to use and I'll upload it to "static.mrpt.org".

**Comment by jlblancoc** [comment]
Ah! I forgot: obviously, you must handle command line arguments in the app for all this to work :-)

It's also useful for lovers of the terminal, as we are all here: we can open a simplemap with: 

```bash
robot-map-gui  mymap.simplemap
```

Also: the equivalent in Windows to all those steps I listed above for Debian is adding one more line [here](https://github.com/MRPT/mrpt/blob/master/cmakemodules/script_setup_cpack.cmake#L85). At least, add it for ".simplemap", since I'm not sure if Windows will drive mad if one uses ".simplemap.gz" as an extension filter... (sigh).

**Comment by jlblancoc** [comment]
Just probably a minor warning: when robot-gui-map is launched from the terminal, one can see this message:

```
QLayout: Attempting to add QLayout "" to CConfigWidget "m_configWidget", which already has a layout
```

**Comment by jlblancoc** [comment]
Thinking of first-time users: if we open a simplemap and, by default, the "configuration" panel is hidden, it might be frustrating to see that the map apparently has been "loaded"... but nothing is shown! :-)

My recommendations: 
* either show (vs. hide) the "configuration" panel by default at start-up, and/or
* show a text message in the area used for maps, that says something like "No metric map defined: please use blah blah... to add one". 

Don't you agree?

**Comment by jlblancoc** [comment]
When I click "apply" (in the configuration pane), I could swear that the CGridPlaneXY gets "stronger"... as if a new opengl object is added (i.e. a duplicate object!) and since each one has an alpha<1.0, the transparencies "add-up".

You can use the method `COpenGLScene::dumpListOfObjects()` to dump the objects to a text file and help you debug this and check for duplicated OpenGL objects.

**Comment by jlblancoc** [comment]
When one selects an observation, it is correctly rendered but if one rotates the 3D view or refresh it somehow, it disappears. I guess it's because you directly call render() of the CSetOfObjects that represent the observation instead of adding it to the scene, right?? 

If you do this to avoid modifying the scene, note that you can use a fixed name for any CRenderizable object via setName() (e.g. `setName("temp_selected_obs")`) so it's easy to remove it afterwards... 

(Update): I realized that this bug only happens when "show all observations" is checked. I hope this helps identifying the problem.

**Comment by jlblancoc** [comment]
I still have the problem of the program using the numeric locale of Spain instead of the "C" standard (!!)...

Please see this example (notice the "," for decimal separator): 

![screenshot from 2017-08-19 02-35-12](https://user-images.githubusercontent.com/5497818/29481876-14bd6ad8-8487-11e7-81f3-9379a6ba8f36.png)

It's fixed by invoking the program like:

```bash
LC_NUMERIC="C" bin/robot-map-gui
``` 

but obviously that's not comfortable to type all the time :-)

Did you tried to fix it and whatever you did failed? Or it's a known task still in the "to-do list"?

BTW: the "re-open" submenu is also still a useful feature that would be great to have before the end of GSoC. 

Despite the long list of comments above... great work! You have contributed a really useful piece of software to MRPT :+1: :+1:

**Comment by jlblancoc** [comment]
One more thing: I tried to "load config" from a file, and the app abruptly exited with an exception: 

* First, wrap "dangerous" areas (mainly... almost all the methods that load/save/update/build/... a map) with a `try{...}catch(...)`  block, such that exceptions could be shown in a Qt messagebox instead of aborting the program. Look for WX_START_TRY / WX_END_TRY for a way to handle this, which you could even also add as another macro to a new file `gui/QtUtils.h` if you see it might be useful. 

* Second: the app should ask the user which is the **section name** that will be used to load the configuration from the config file... by default, it should be "MappingApplication" for backwards compatibility, but users might want to change it. 

In my case, I think that the crash was related with the use of the Spanish locale... any number like "0.5" is considered to be "0"... (sigh). 
PD: Yes, I can confirm that the crash doesn't happen with using the "C" locale. However, the need to gets protected about an exception crashing the program is still a real need!

**Comment by jlblancoc** [comment]
I also noticed that the "move" operation works perfectly, thanks!

However... it seems to be enabled by default, right? A map is, in theory, hard to build so it would be a pity that someone accidentally moves a robot pose, ruining a part of the map.

What do you think about this? Would it be worth adding some check box/or-alike to enable "moving"? (it should be disabled by default).

**Comment by jlblancoc** [comment]
Final point for tonight, I promise ;-)

All `CMetricMap`s, including `CMultiMetricMap`, have a method [`saveMetricMapRepresentationToFile()`](http://mrpt.ual.es/reference/devel/classmrpt_1_1maps_1_1_c_multi_metric_map.html#aef74bee8ef6315012ce967dc11fc284e). Please, add a new menu item under "File" -> "Save metric maps representations..." that asks the user for a file name to be used as the prefix passed to saveMetricMapRepresentationToFile() for all the metric maps, so they get saved to disk in the format that each metric map itself considers. 

By the way: exporting as "binary" object would be also useful in many situations: another item "File" -> "Save metric map in binary format..." would call to: 

```cpp
CMetricMap * my_currently_visible_map;
// ...
mrpt::utils::CFileGZOutputStream fil(file_name);
fil << *my_currently_visible_map;
```

**Comment by jlblancoc** [comment]
@LisGein I saw you've been making progress on these topics above... here are my tests: 
* "C" locale: It still showed my annoying Spanish "," for decimals... I imagined what may be causing this, by similarity with similar problems with wxWidgets. Please, apply this patch that fixes it: 

```diff
--- a/apps/robot-map-gui/robot-map-gui_main.cpp
+++ b/apps/robot-map-gui/robot-map-gui_main.cpp
@@ -21,7 +21,6 @@
 
 int main(int argc, char** argv)
 {
-       setlocale(LC_NUMERIC, "C");
        QLocale::setDefault(QLocale::C);
 
        QApplication app(argc, argv);
@@ -29,6 +28,8 @@ int main(int argc, char** argv)
        QApplication::setOrganizationDomain("mrpt.org");
        QApplication::setApplicationName("robot-map-gui");
 
+        setlocale(LC_NUMERIC, "C");
+
        QCommandLineParser parser;
        parser.parse(QApplication::arguments());
        const QStringList args = parser.positionalArguments();
```

* Opening a file via CLI: Works! :ok_hand: 
However, if one passes a file that doesn't exist (yes... :smiling_imp: ), an exception message is dumped to console only, without any warning in the GUI. This would probably be also fixed by adding the equivalent to WX_START_TRY/WX_END_TRY for Qt that was mentioned above.

* Recent files:  Works! :ok_hand: 


Keep up the hard work!

**Comment by LisGein** [comment]
Hi! 
Thank you for good review!


 > QLayout: Attempting to add QLayout "" to CConfigWidget "m_configWidget", which already has a layout

I think it's Qt bug because it problem in *.ui file and I can't change it.


I am doing a warning in the GUI about "bad" file and text message in the area of the maps when don't open simplemap and/or config file.  
Also I'm doing saveMetricMapRepresentationToFile().

**Comment by LisGein** [comment]
Hi! I've done implement "About", added save any metric map in binary format and representation to file. Also, I've fixed locale in Qt, added a warning window(for bad *.simplemap), added the opening the application in the console.

Now, I wanna add graph-slam to my program.
I found the function
`CSetOfObjects::Ptr graph_visualize(const GRAPH_T& g, const mrpt::utils::TParametersDouble& extra_params)`

It gets the class like CNetworkOfPoses. How can I get CNetworkOfPoses? 
Or can you give me other information about this task?

**Comment by jlblancoc** [comment]
Hi,

I saw your latest changes regarding the refactoring of parameters <=> GUI... good. Anything that makes easier maintaining the code it's welcome!

On graphslam: it's actually much more complicated... Obtaining the pose constraints in CNetworkOfPoses requires trying to "align" each CSensoryFrame with all its neighbors. Making it general for all kind of observations would be too complex and certainly out of the reach for the time you have at this point. 
But you could give it a try to only implement it for 2D laser scans (which is one fairly common sensor in robotics). For that, please give it a quick read to [this paper](http://domino.informatik.uni-freiburg.de/teaching/ws11/robotics2/pdfs/ls-slam-tutorial.pdf) just to get the global idea (skip the maths!). 

Then, assuming that the `.simplemap` is already a "good map" (with all loop closures already correctly detected and closed, etc.) then graph-slam could be used to trying to "refine" the map. 
This can be done by executing ICP between **every pair of nearby CSensoryFrames** (you should look at the corresponding global pose for each CSensoryFrame). 
Then, build one temporary pointsmap for each CSensoryFrame and execute ICP against the two maps.
See [this example code](https://github.com/MRPT/mrpt/blob/master/samples/icp/test.cpp#L205). Ideally, the "initial guess" passed to ICP should be the relative pose between the two CSensoryFrames. You can use the C++ operator `-` applied to a CPose2D or CPose3D (like `p2 - p1`) to compute the relative pose or "p2" as seen from "p1". 
Then, the output from ICP, only if the quality output from it is high enough, becomes **one edge** in the CNetworkOfPoses. 
When all edges have been added, we can optimize the graph, and obtain the new, in theory improved, global poses for each robot pose, and overwrite the poses in the original .simplemap.  
See [this example](https://github.com/MRPT/mrpt/blob/master/libs/graphslam/src/graph_slam_levmarq_unittest.cpp#L29) for how to define a CNetworkOfPoses and optimize it.


As I told above... all this is fairly complex specially if you never faced relative poses or graphslam before... but just give it a try and let us know if you find some difficulty. 

PS: The only remaining "issue" I can see with the curent GUI is that some "panels" of parameters have a relatively large "height", so depending on the screen resolution this may become an issue... would it be easy to embed the largest panels inside a `QScrollArea` or alike?

PS2: When you feel all the tasks discussed above (except the graph-slam stuff) is ready, feel free of opening a PR against MRPT/master. Later you would be able to open new ones if that's needed, but the sooner I can take a look at the many merge conflicts that apparently we have, the better...

**Comment by LisGein** [comment]
Hi! 
I created PR for the main part of the work.
I moved changes related to reflection to the separate branch. I think that changes may cause some discussions. After finish this task I will create new PR.

**Comment by LisGein** [comment]
Also, I did the [video](https://youtu.be/YgGcNvrxZI0).

**Comment by LisGein** [comment]
Please, review my summary works information:
https://gist.github.com/LisGein/1ad3951ecc8a67d4915a59ae042b9f45

**Comment by jlblancoc** [comment]
Great, Alice!

I like the video👍👍

We'll review the pr in the next days. We mentors have until first week on
September to make our final report, so expect some comments before that
date.

Congratulations for the great work... your app is fairly better and more
appealing than its 10-years old Windows-only predecessor! I'm glad of
having this functionality back into mrpt after so many years ☺

**Comment by LisGein** [comment]
Thank you!
Is it correct to paste this [link](github.com/LisGein/1ad3951ecc8a67d4915a59ae042b9f45) in  "Provide a link to the work you did during GSoC 2017"?

**Comment by LisGein** [comment]
Hi, @jlblancoc ! I'm a little bit worried about deadline(1 day left). Google says "You should share your link with your mentor BEFORE submitting your evaluation to make sure it's okay."
[My summary works.](https://gist.github.com/LisGein/1ad3951ecc8a67d4915a59ae042b9f45)

I added commits for your comments. 
I wanna do synchronizing select poses on this week
Also, I wanna do changing poses orientation in the future.

**Comment by jlblancoc** [comment]
Hi Alice, 

No worries for the deadline: your summary is perfect. Go on and use it to do your submission. 

We can continue and discuss on more "TO-DO"s later. Will re-check the PR and merge it. When you add new functionality, or we come back to you if we find a bug, etc. you will be able to open new PRs, naturally ;-)

Again, thanks for your great contributions!

**Comment by LisGein** [comment]
Of course. 
I am having a vacation until 18th September. After it, I wanna continue to add new commits :)

It internship was a really great opportunity for me! Thanks a lot!!!

**Comment by jlblancoc** [comment]
It was a pleasure for us too, thank *you*!
Enjoy your vacation :+1: 

I'll close now this ticket, since the PRs are already merged and the app is in a "functional" state and we are approaching the end of GSOC.

We can continue discussion here about the pending topics after 19th September, anyway. 
Looking forward to that!
